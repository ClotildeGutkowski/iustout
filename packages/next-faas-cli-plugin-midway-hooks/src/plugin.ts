import { BasePlugin } from '@midwayjs/fcli-command-core'
import { resolve } from 'path'
import {
  hintConfig,
  helper,
  getFunctionsMeta,
  clearRoutes,
  MidwayHooksFunctionStructure,
} from '@midwayjs/next-hooks-compiler'
import { debug, argsPath } from './util'
import { WatcherConfig, HooksWatcher } from './watcher'
import { transform, EventStructureType } from '@midwayjs/serverless-spec-builder'
import type { SpecStructureWithGateway } from '@midwayjs/hooks-shared'

export class MidwayHooksPlugin extends BasePlugin {
  spec: SpecStructureWithGateway

  get gateway() {
    if (!this.spec) {
      this.spec = transform(resolve(this.root, 'f.yml'))
    }

    return this.spec.apiGateway?.type ?? 'http'
  }

  get root() {
    return this.core.config.servicePath
  }

  get apis() {
    return resolve(this.root, 'src/apis')
  }

  get isSkipTsBuild() {
    return this.getStore('skipTsBuild', 'global')
  }

  private static init = false
  private static isCompiled = false

  hooks = {
    'before:invoke:compile': async () => {
      this.beforeCompile()
    },
    'after:invoke:emit': async () => {
      await this.afterCompile()
    },
    'before:package:compile': async () => {
      this.beforeCompile()
    },
    'after:package:emit': async () => {
      await this.afterCompile()
    },
  }

  beforeCompile() {
    if (this.isSkipTsBuild) {
      return
    }

    if (!MidwayHooksPlugin.init) {
      debug('hintConfig: %O', hintConfig)
      MidwayHooksPlugin.init = true
    }

    debug('beforeCompile')

    if ((this.core.service as any).functionsRule) {
      helper.rules = (this.core.service as any).functionsRule
      helper.root = resolve(this.root, 'src')
    } else {
      helper.root = this.apis
    }

    clearRoutes()

    this.setStore('mwccHintConfig', hintConfig, true)
  }

  async afterCompile() {
    if (this.isSkipTsBuild) {
      return
    }

    debug('afterCompile')

    const functions = getFunctionsMeta()
    for (const func of Object.values(functions)) {
      func.gatewayConfig.meta.gateway = this.gateway
      func.argsPath = this.argsPath
      func.events = [this.createEventsByGateway(func)]
    }

    debug('faas decorator function: %O', this.core.service.functions)
    debug('hooks function: %s', JSON.stringify(functions, null, 2))

    this.core.service.functions = Object.assign(this.core.service.functions, functions)

    if (!MidwayHooksPlugin.isCompiled && process.env.NODE_ENV !== 'production') {
      startWatcher({
        root: this.root,
        apis: this.apis,
      })
      MidwayHooksPlugin.isCompiled = true
    }
  }

  protected get argsPath() {
    return argsPath[this.gateway]
  }

  protected createEventsByGateway(func: MidwayHooksFunctionStructure): EventStructureType {
    const config = func.gatewayConfig
    /**
     * HTTP Case
     */
    return {
      http: {
        method: config.method,
        path: config.url,
      },
    }
  }
}

async function startWatcher(config: WatcherConfig) {
  const watcher = new HooksWatcher(config)

  while (true) {
    await watcher.getNext()
  }
}
