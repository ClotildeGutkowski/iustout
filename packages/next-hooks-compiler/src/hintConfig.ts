export const hintConfig = {
  features: {
    tsc: {
      transformers: [
        {
          name: require.resolve('./plugin/ref-to-bind'),
        },
        {
          name: require.resolve('./plugin/anonymous-function-to-named'),
        },
        {
          name: require.resolve('./plugin/create-lambda-ctx'),
        },
        {
          name: require.resolve('./plugin/sourcefile-logger'),
        },
      ],
    },
  },
}
