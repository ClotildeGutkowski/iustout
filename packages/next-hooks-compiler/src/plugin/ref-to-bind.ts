import ts from 'typescript'
import { TransformationContext, template } from '@midwayjs/mwcc'
import { MidwayHooksPackage, HooksMethodNamespace, BuiltinHooks, ContextBind } from '../const'
import { BuiltinHooksError } from '../errors/BuiltinHooks'
import { isEmpty } from 'lodash'
import {
  isHookName,
  closetAncestor,
  getSourceFilePath,
  isLambdaOrHook,
  isAncestor,
  getTopLevelNode,
  isInsideLambdaOrHook,
} from '../util'
import { helper } from '../helper'
import { InvalidReferenceError } from '../errors/InvalidReference'
import { debug } from '../util'

export default {
  transform(ctx: TransformationContext) {
    return {
      Identifier(node: ts.Identifier) {
        if (
          closetAncestor(node, ts.SyntaxKind.ImportDeclaration) ||
          closetAncestor(node, ts.SyntaxKind.BindingElement)
        ) {
          return node
        }

        let declarations: ts.Node[]
        try {
          declarations = ctx.resolveDeclarations(node)
        } catch (error) {
          debug('ctx.resolveDeclarations. Error: %s, Identifier: %s', error.message, node.getText())
          return node
        }

        if (!isEmpty(declarations)) {
          if (ts.isBindingElement(declarations[0])) {
            return node
          }
        }

        const importNames = ctx.resolveImportedNames(node)

        if (isEmpty(importNames)) {
          return processReference(node, ctx)
        }

        const [{ moduleId }] = importNames
        return processImportNames(node, moduleId, ctx)
      },
    }
  },
}

function processReference(node: ts.Identifier, ctx: TransformationContext) {
  const isRef = isHookName(node.text) || helper.isLambdaFile(getSourceFilePath(node))

  if (!isRef) {
    return node
  }

  const declarations = ctx
    .resolveDeclarations(node)
    .filter((declaration) => helper.isProjectFile(getSourceFilePath(declaration)))

  if (isEmpty(declarations)) {
    return node
  }

  const [declaration] = declarations

  if (isLambdaOrHook(getTopLevelNode(declaration), declaration) && !isAncestor(node, declaration)) {
    if (!isInsideLambdaOrHook(node)) {
      throw new InvalidReferenceError(node)
    }

    return compileRefToBind(node)
  }

  return node
}

function processImportNames(node: ts.Identifier, moduleId: string, ctx: TransformationContext) {
  if (!isLambdaOrHookImports(node, moduleId, ctx)) {
    return node
  }

  if (!isInsideLambdaOrHook(node)) {
    throw new InvalidReferenceError(node)
  }

  // 不处理 Type 引用
  if (closetAncestor(node, ts.SyntaxKind.TypeQuery) || closetAncestor(node, ts.SyntaxKind.TypeReference)) {
    return node
  }

  if (moduleId === MidwayHooksPackage) {
    return compileBuiltinHooks(node)
  } else {
    return compileRefToBind(node)
  }
}

function isLambdaOrHookImports(node: ts.Identifier, moduleId: string, ctx: TransformationContext) {
  // lodash
  if (!moduleId.startsWith('.')) {
    // @midwayjs/hooks 内置 Hooks
    if (moduleId === MidwayHooksPackage) {
      // 忽略 @MidwayHooks 装饰器
      return true
    }

    return false
  }

  const declarations = ctx.resolveDeclarations(node)
  if (isEmpty(declarations)) {
    return false
  }

  const [definition] = declarations
  return closetAncestor(definition, ts.SyntaxKind.ImportDeclaration) && isHookName(node.getText())
}

/**
 * useQuery => useQuery.bind($lambda)
 */
function compileRefToBind(identifier: ts.Identifier) {
  const tpl = template(`HOOK.${ContextBind}`)({ HOOK: identifier })[0] as ts.ExpressionStatement
  return tpl.expression
}

/**
 * useContext => $lambda.ctx.hooks.useContext
 */
function compileBuiltinHooks(identifier: ts.Identifier) {
  const hook = identifier.getText()

  if (!BuiltinHooks.includes(hook)) {
    throw new BuiltinHooksError(identifier)
  }

  const tpl = template(`${HooksMethodNamespace}.HOOK`)({ HOOK: identifier })[0] as ts.ExpressionStatement
  return tpl.expression
}
