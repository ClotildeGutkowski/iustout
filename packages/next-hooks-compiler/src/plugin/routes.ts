import { GatewayConfig } from './gateway/interface'
import type { FunctionStructure } from '@midwayjs/serverless-spec-builder'
import type { Dictionary } from 'lodash'

type SourceFilePath = string

const map = new Map<SourceFilePath, MidwayHooksFunctionStructure[]>()

export function addRoute(sourceFilePath: SourceFilePath, gateway: MidwayHooksFunctionStructure) {
  if (!map.has(sourceFilePath)) {
    map.set(sourceFilePath, [])
  }

  const routes = map.get(sourceFilePath)

  const idx = routes.findIndex((config) => config.handler === gateway.handler)
  if (idx !== -1) {
    routes.splice(idx, 1)
  }

  routes.push(gateway)
}

export function getRoutes(sourceFilePath: SourceFilePath) {
  return map.get(sourceFilePath) ?? []
}

export function clearRoutes() {
  map.clear()
}

export function getFunctionsMeta(): Dictionary<MidwayHooksFunctionStructure> {
  const functions: Dictionary<MidwayHooksFunctionStructure> = {}

  map.forEach((configs) => {
    for (const config of configs) {
      functions[config.gatewayConfig.handler] = config
    }
  })

  return functions
}

export interface MidwayHooksFunctionStructure extends FunctionStructure {
  handler: string
  sourceFilePath?: string
  exportFunction?: string
  isFunctional?: boolean
  argsPath?: string
  gatewayConfig: Partial<GatewayConfig>
}
