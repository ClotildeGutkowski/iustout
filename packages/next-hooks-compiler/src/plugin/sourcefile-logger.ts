import ts from 'typescript'
import { TransformationContext } from '@midwayjs/mwcc'
import { getSourceFilePath } from '../util'
import { debug } from '../util'
import { helper } from '../helper'
import { relative } from 'upath'

export default {
  transform(ctx: TransformationContext) {
    return {
      SourceFile(node: ts.SourceFile) {
        const sourceFilePath = getSourceFilePath(node)
        debug('SourceFile %s', relative(helper.source, sourceFilePath))
        return node
      },
    }
  },
}
