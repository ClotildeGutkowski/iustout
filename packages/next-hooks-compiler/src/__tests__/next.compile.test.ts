import path from 'upath'
import { hintConfig } from '../hintConfig'
import { compileHooks } from '../index'
import globby from 'globby'
import fse from 'fs-extra'
import { getFunctionsMeta } from '../plugin/routes'

describe('NeXT Hooks Compiler', () => {
  const fixture = path.resolve(__dirname, './fixtures/hook')
  const source = path.resolve(fixture, 'src')
  const dist = path.resolve(fixture, 'dist')

  beforeAll(async () => {
    await compileHooks(fixture, hintConfig)
  })

  const files = globby.sync(source)

  for (const file of files) {
    const relative = path.relative(source, file)
    const target = path.resolve(dist, relative).replace('.ts', '.js')

    it(relative, async () => {
      const content = await fse.readFile(file, 'utf-8')
      const compiled = await fse.readFile(target, 'utf-8')

      expect(`
// 源文件

${content}

// 编译结果

${compiled}
      `).toMatchSnapshot()
    })
  }

  it('路由信息应该生成正确', () => {
    expect(getFunctionsMeta()).toMatchSnapshot()
  })
})
