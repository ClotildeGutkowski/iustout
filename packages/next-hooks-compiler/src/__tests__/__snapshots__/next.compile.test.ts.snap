// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NeXT Hooks Compiler hooks/arrow.ts 1`] = `
"
// 源文件

export const useDemo = (name: string) => {}

export const useArrow = (name: string) => {
  useDemo(name + '666')
}

export const useAsyncArrow = async (name: string) => {
  await useDemo(name + '666')
}


// 编译结果

\\"use strict\\";
Object.defineProperty(exports, \\"__esModule\\", { value: true });
exports.useAsyncArrow = exports.useArrow = exports.useDemo = void 0;
exports.useDemo = function useDemo(name) { const $lambda = this; };
exports.useArrow = function useArrow(name) { const $lambda = this; useDemo.bind($lambda)(name + '666'); };
exports.useAsyncArrow = async function useAsyncArrow(name) { const $lambda = this; await useDemo.bind($lambda)(name + '666'); };
//# sourceMappingURL=arrow.js.map
      "
`;

exports[`NeXT Hooks Compiler hooks/arrow-default-export.ts 1`] = `
"
// 源文件

export default async (name: string) => {}


// 编译结果

\\"use strict\\";
Object.defineProperty(exports, \\"__esModule\\", { value: true });
exports.default = async (name) => { };
//# sourceMappingURL=arrow-default-export.js.map
      "
`;

exports[`NeXT Hooks Compiler hooks/builtin-hooks.ts 1`] = `
"
// 源文件

import { useContext, useInject, useConfig, useLogger, usePlugin } from '@midwayjs/hooks'

export function useBuiltinHooks() {
  useConfig()
  useLogger()
  useContext()
  useInject('')
  usePlugin('')
}


// 编译结果

\\"use strict\\";
Object.defineProperty(exports, \\"__esModule\\", { value: true });
exports.useBuiltinHooks = void 0;
const hooks_1 = require(\\"@midwayjs/hooks\\");
function useBuiltinHooks() {
    const $lambda = this;
    $lambda.ctx.hooks.useConfig();
    $lambda.ctx.hooks.useLogger();
    $lambda.ctx.hooks.useContext();
    $lambda.ctx.hooks.useInject('');
    $lambda.ctx.hooks.usePlugin('');
}
exports.useBuiltinHooks = useBuiltinHooks;
//# sourceMappingURL=builtin-hooks.js.map
      "
`;

exports[`NeXT Hooks Compiler hooks/func-default-export.ts 1`] = `
"
// 源文件

export default function (name: string) {}


// 编译结果

\\"use strict\\";
Object.defineProperty(exports, \\"__esModule\\", { value: true });
function default_1(name) { }
exports.default = default_1;
//# sourceMappingURL=func-default-export.js.map
      "
`;

exports[`NeXT Hooks Compiler hooks/import-hook.ts 1`] = `
"
// 源文件

import { useDemo } from './useDemo'
import { useContext } from '@midwayjs/hooks'

export function useQuery(name: any) {
  useContext()
  useDemo(name)
  console.log(useFakeHooks)
}

export class ThisNotHooks {}

export const useFakeHooks = ''


// 编译结果

\\"use strict\\";
Object.defineProperty(exports, \\"__esModule\\", { value: true });
exports.useFakeHooks = exports.ThisNotHooks = exports.useQuery = void 0;
const useDemo_1 = require(\\"./useDemo\\");
const hooks_1 = require(\\"@midwayjs/hooks\\");
function useQuery(name) {
    const $lambda = this;
    $lambda.ctx.hooks.useContext();
    useDemo_1.useDemo.bind($lambda)(name);
    console.log(useFakeHooks);
}
exports.useQuery = useQuery;
class ThisNotHooks {
}
exports.ThisNotHooks = ThisNotHooks;
exports.useFakeHooks = '';
//# sourceMappingURL=import-hook.js.map
      "
`;

exports[`NeXT Hooks Compiler hooks/import-namespace.ts 1`] = `
"
// 源文件

import * as hooks from './import-hook'

export function useNamespace() {
  const n = new hooks.ThisNotHooks()
  console.log(hooks.useFakeHooks)
  return hooks.useQuery(n)
}


// 编译结果

\\"use strict\\";
Object.defineProperty(exports, \\"__esModule\\", { value: true });
exports.useNamespace = void 0;
const hooks = require(\\"./import-hook\\");
function useNamespace() {
    const $lambda = this;
    const n = new hooks.ThisNotHooks();
    console.log(hooks.useFakeHooks);
    return hooks.useQuery.bind($lambda)(n);
}
exports.useNamespace = useNamespace;
//# sourceMappingURL=import-namespace.js.map
      "
`;

exports[`NeXT Hooks Compiler hooks/useDemo.ts 1`] = `
"
// 源文件

import { useContext } from '@midwayjs/hooks'

export function useDemo(name: string) {
  useContext()
}


// 编译结果

\\"use strict\\";
Object.defineProperty(exports, \\"__esModule\\", { value: true });
exports.useDemo = void 0;
const hooks_1 = require(\\"@midwayjs/hooks\\");
function useDemo(name) { const $lambda = this; $lambda.ctx.hooks.useContext(); }
exports.useDemo = useDemo;
//# sourceMappingURL=useDemo.js.map
      "
`;

exports[`NeXT Hooks Compiler lambda/arrow-hook-ref-arrow-hook.ts 1`] = `
"
// 源文件

export const useQuery = (id: string) => {
  return ''
}

export const useInner = async () => {
  useQuery('123')
  const inner = () => {
    console.log(useQuery('123'))
  }
  inner()
}


// 编译结果

\\"use strict\\";
Object.defineProperty(exports, \\"__esModule\\", { value: true });
exports.useInner = exports.useQuery = void 0;
exports.useQuery = function useQuery(id) { const $lambda = this; return ''; };
exports.useInner = async function useInner() {
    const $lambda = this;
    useQuery.bind($lambda)('123');
    const inner = () => {
        console.log(useQuery.bind($lambda)('123'));
    };
    inner();
};
//# sourceMappingURL=arrow-hook-ref-arrow-hook.js.map
      "
`;

exports[`NeXT Hooks Compiler lambda/export-arrow.ts 1`] = `
"
// 源文件

export const hello = (name: string) => {
  console.log(name)
}


// 编译结果

\\"use strict\\";
Object.defineProperty(exports, \\"__esModule\\", { value: true });
exports.hello = void 0;
exports.hello = function hello(name) { const $lambda = this; console.log(name); };
//# sourceMappingURL=export-arrow.js.map
      "
`;

exports[`NeXT Hooks Compiler lambda/export-default-anonymous.ts 1`] = `
"
// 源文件

export default () => {}


// 编译结果

\\"use strict\\";
Object.defineProperty(exports, \\"__esModule\\", { value: true });
function $default() { const $lambda = this; }
exports.default = $default;
//# sourceMappingURL=export-default-anonymous.js.map
      "
`;

exports[`NeXT Hooks Compiler lambda/export-default-func.ts 1`] = `
"
// 源文件

export default function (name: string) {}


// 编译结果

\\"use strict\\";
Object.defineProperty(exports, \\"__esModule\\", { value: true });
function $default(name) { const $lambda = this; }
exports.default = $default;
//# sourceMappingURL=export-default-func.js.map
      "
`;

exports[`NeXT Hooks Compiler lambda/export-default-named.ts 1`] = `
"
// 源文件

export default function named(name: string) {}


// 编译结果

\\"use strict\\";
Object.defineProperty(exports, \\"__esModule\\", { value: true });
function named(name) { const $lambda = this; }
exports.default = named;
//# sourceMappingURL=export-default-named.js.map
      "
`;

exports[`NeXT Hooks Compiler lambda/hook-ref-hook.ts 1`] = `
"
// 源文件

export function useQuery(id: string) {
  return ''
}

export async function useInner() {
  useQuery('123')
  const inner = () => {
    console.log(useQuery('123'))
  }
  inner()
}


// 编译结果

\\"use strict\\";
Object.defineProperty(exports, \\"__esModule\\", { value: true });
exports.useInner = exports.useQuery = void 0;
function useQuery(id) { const $lambda = this; return ''; }
exports.useQuery = useQuery;
async function useInner() {
    const $lambda = this;
    useQuery.bind($lambda)('123');
    const inner = () => {
        console.log(useQuery.bind($lambda)('123'));
    };
    inner();
}
exports.useInner = useInner;
//# sourceMappingURL=hook-ref-hook.js.map
      "
`;

exports[`NeXT Hooks Compiler lambda/hook-ref-lambda.ts 1`] = `
"
// 源文件

export function hello() {
  normal()
}

function normal() {}

function useDemo(name: string) {
  return hello()
}


// 编译结果

\\"use strict\\";
Object.defineProperty(exports, \\"__esModule\\", { value: true });
exports.hello = void 0;
function hello() { const $lambda = this; normal(); }
exports.hello = hello;
function normal() { }
function useDemo(name) { const $lambda = this; return hello.bind($lambda)(); }
//# sourceMappingURL=hook-ref-lambda.js.map
      "
`;

exports[`NeXT Hooks Compiler lambda/import-hook.ts 1`] = `
"
// 源文件

import { useDemo } from '../hooks/useDemo'

export function hello() {
  useDemo('666')
}


// 编译结果

\\"use strict\\";
Object.defineProperty(exports, \\"__esModule\\", { value: true });
exports.hello = void 0;
const useDemo_1 = require(\\"../hooks/useDemo\\");
function hello() { const $lambda = this; useDemo_1.useDemo.bind($lambda)('666'); }
exports.hello = hello;
//# sourceMappingURL=import-hook.js.map
      "
`;

exports[`NeXT Hooks Compiler lambda/index.ts 1`] = `
"
// 源文件

export function hello() {}


// 编译结果

\\"use strict\\";
Object.defineProperty(exports, \\"__esModule\\", { value: true });
exports.hello = void 0;
function hello() { const $lambda = this; }
exports.hello = hello;
//# sourceMappingURL=index.js.map
      "
`;

exports[`NeXT Hooks Compiler lambda/lambda-ref-hook.ts 1`] = `
"
// 源文件

export function hello() {
  useDemo('666')
  normal()
}

function normal() {}

function useDemo(name: string) {}


// 编译结果

\\"use strict\\";
Object.defineProperty(exports, \\"__esModule\\", { value: true });
exports.hello = void 0;
function hello() {
    const $lambda = this;
    useDemo.bind($lambda)('666');
    normal();
}
exports.hello = hello;
function normal() { }
function useDemo(name) { const $lambda = this; }
//# sourceMappingURL=lambda-ref-hook.js.map
      "
`;

exports[`NeXT Hooks Compiler lambda/lambda-ref-lambda.ts 1`] = `
"
// 源文件

export function hello(name: string, value: string) {}

const hello3 = (name: string) => {}

export function hello2(name: string) {
  hello(name, '1')
  hello3(name)
}


// 编译结果

\\"use strict\\";
Object.defineProperty(exports, \\"__esModule\\", { value: true });
exports.hello2 = exports.hello = void 0;
function hello(name, value) { const $lambda = this; }
exports.hello = hello;
const hello3 = (name) => { };
function hello2(name) {
    const $lambda = this;
    hello.bind($lambda)(name, '1');
    hello3(name);
}
exports.hello2 = hello2;
//# sourceMappingURL=lambda-ref-lambda.js.map
      "
`;

exports[`NeXT Hooks Compiler lambda/normal-function.ts 1`] = `
"
// 源文件

export function hello() {
  normal()
}

function normal() {}


// 编译结果

\\"use strict\\";
Object.defineProperty(exports, \\"__esModule\\", { value: true });
exports.hello = void 0;
function hello() { const $lambda = this; normal(); }
exports.hello = hello;
function normal() { }
//# sourceMappingURL=normal-function.js.map
      "
`;

exports[`NeXT Hooks Compiler 路由信息应该生成正确 1`] = `
Object {
  "arrow-hook-ref-arrow-hook-useinner": Object {
    "argsPath": "",
    "exportFunction": "useInner",
    "gatewayConfig": Object {
      "handler": "arrow-hook-ref-arrow-hook-useinner",
      "meta": Object {
        "functionName": "useInner",
      },
      "method": "GET",
      "url": "/api/arrow-hook-ref-arrow-hook/_useInner",
    },
    "handler": "arrow-hook-ref-arrow-hook-useinner.handler",
    "isFunctional": true,
    "sourceFilePath": "lambda/arrow-hook-ref-arrow-hook.js",
  },
  "arrow-hook-ref-arrow-hook-usequery": Object {
    "argsPath": "",
    "exportFunction": "useQuery",
    "gatewayConfig": Object {
      "handler": "arrow-hook-ref-arrow-hook-usequery",
      "meta": Object {
        "functionName": "useQuery",
      },
      "method": "POST",
      "url": "/api/arrow-hook-ref-arrow-hook/_useQuery",
    },
    "handler": "arrow-hook-ref-arrow-hook-usequery.handler",
    "isFunctional": true,
    "sourceFilePath": "lambda/arrow-hook-ref-arrow-hook.js",
  },
  "export-arrow-hello": Object {
    "argsPath": "",
    "exportFunction": "hello",
    "gatewayConfig": Object {
      "handler": "export-arrow-hello",
      "meta": Object {
        "functionName": "hello",
      },
      "method": "POST",
      "url": "/api/export-arrow/_hello",
    },
    "handler": "export-arrow-hello.handler",
    "isFunctional": true,
    "sourceFilePath": "lambda/export-arrow.js",
  },
  "export-default-anonymous": Object {
    "argsPath": "",
    "exportFunction": "",
    "gatewayConfig": Object {
      "handler": "export-default-anonymous",
      "meta": Object {
        "functionName": "$default",
      },
      "method": "GET",
      "url": "/api/export-default-anonymous",
    },
    "handler": "export-default-anonymous.handler",
    "isFunctional": true,
    "sourceFilePath": "lambda/export-default-anonymous.js",
  },
  "export-default-func": Object {
    "argsPath": "",
    "exportFunction": "",
    "gatewayConfig": Object {
      "handler": "export-default-func",
      "meta": Object {
        "functionName": "$default",
      },
      "method": "POST",
      "url": "/api/export-default-func",
    },
    "handler": "export-default-func.handler",
    "isFunctional": true,
    "sourceFilePath": "lambda/export-default-func.js",
  },
  "export-default-named": Object {
    "argsPath": "",
    "exportFunction": "",
    "gatewayConfig": Object {
      "handler": "export-default-named",
      "meta": Object {
        "functionName": "named",
      },
      "method": "POST",
      "url": "/api/export-default-named",
    },
    "handler": "export-default-named.handler",
    "isFunctional": true,
    "sourceFilePath": "lambda/export-default-named.js",
  },
  "hook-ref-hook-useinner": Object {
    "argsPath": "",
    "exportFunction": "useInner",
    "gatewayConfig": Object {
      "handler": "hook-ref-hook-useinner",
      "meta": Object {
        "functionName": "useInner",
      },
      "method": "GET",
      "url": "/api/hook-ref-hook/_useInner",
    },
    "handler": "hook-ref-hook-useinner.handler",
    "isFunctional": true,
    "sourceFilePath": "lambda/hook-ref-hook.js",
  },
  "hook-ref-hook-usequery": Object {
    "argsPath": "",
    "exportFunction": "useQuery",
    "gatewayConfig": Object {
      "handler": "hook-ref-hook-usequery",
      "meta": Object {
        "functionName": "useQuery",
      },
      "method": "POST",
      "url": "/api/hook-ref-hook/_useQuery",
    },
    "handler": "hook-ref-hook-usequery.handler",
    "isFunctional": true,
    "sourceFilePath": "lambda/hook-ref-hook.js",
  },
  "hook-ref-lambda-hello": Object {
    "argsPath": "",
    "exportFunction": "hello",
    "gatewayConfig": Object {
      "handler": "hook-ref-lambda-hello",
      "meta": Object {
        "functionName": "hello",
      },
      "method": "GET",
      "url": "/api/hook-ref-lambda/_hello",
    },
    "handler": "hook-ref-lambda-hello.handler",
    "isFunctional": true,
    "sourceFilePath": "lambda/hook-ref-lambda.js",
  },
  "import-hook-hello": Object {
    "argsPath": "",
    "exportFunction": "hello",
    "gatewayConfig": Object {
      "handler": "import-hook-hello",
      "meta": Object {
        "functionName": "hello",
      },
      "method": "GET",
      "url": "/api/import-hook/_hello",
    },
    "handler": "import-hook-hello.handler",
    "isFunctional": true,
    "sourceFilePath": "lambda/import-hook.js",
  },
  "index-hello": Object {
    "argsPath": "",
    "exportFunction": "hello",
    "gatewayConfig": Object {
      "handler": "index-hello",
      "meta": Object {
        "functionName": "hello",
      },
      "method": "GET",
      "url": "/api/_hello",
    },
    "handler": "index-hello.handler",
    "isFunctional": true,
    "sourceFilePath": "lambda/index.js",
  },
  "lambda-ref-hook-hello": Object {
    "argsPath": "",
    "exportFunction": "hello",
    "gatewayConfig": Object {
      "handler": "lambda-ref-hook-hello",
      "meta": Object {
        "functionName": "hello",
      },
      "method": "GET",
      "url": "/api/lambda-ref-hook/_hello",
    },
    "handler": "lambda-ref-hook-hello.handler",
    "isFunctional": true,
    "sourceFilePath": "lambda/lambda-ref-hook.js",
  },
  "lambda-ref-lambda-hello": Object {
    "argsPath": "",
    "exportFunction": "hello",
    "gatewayConfig": Object {
      "handler": "lambda-ref-lambda-hello",
      "meta": Object {
        "functionName": "hello",
      },
      "method": "POST",
      "url": "/api/lambda-ref-lambda/_hello",
    },
    "handler": "lambda-ref-lambda-hello.handler",
    "isFunctional": true,
    "sourceFilePath": "lambda/lambda-ref-lambda.js",
  },
  "lambda-ref-lambda-hello2": Object {
    "argsPath": "",
    "exportFunction": "hello2",
    "gatewayConfig": Object {
      "handler": "lambda-ref-lambda-hello2",
      "meta": Object {
        "functionName": "hello2",
      },
      "method": "POST",
      "url": "/api/lambda-ref-lambda/_hello2",
    },
    "handler": "lambda-ref-lambda-hello2.handler",
    "isFunctional": true,
    "sourceFilePath": "lambda/lambda-ref-lambda.js",
  },
  "normal-function-hello": Object {
    "argsPath": "",
    "exportFunction": "hello",
    "gatewayConfig": Object {
      "handler": "normal-function-hello",
      "meta": Object {
        "functionName": "hello",
      },
      "method": "GET",
      "url": "/api/normal-function/_hello",
    },
    "handler": "normal-function-hello.handler",
    "isFunctional": true,
    "sourceFilePath": "lambda/normal-function.js",
  },
}
`;
